<a href="https://colab.research.google.com/github/nelakurthirahul/Data_Projects/blob/main/NYC_Bus_Delays_ETL_Pipeline.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [126]:
import polars as pl
df = pl.read_csv("Bus_Breakdown_and_Delays_20260922.csv",infer_schema_length=0)
df.shape

(1297681, 21)

In [127]:
import requests
url = "https://data.cityofnewyork.us/Transportation/Bus-Breakdown-and-Delays/ez4e-fazm/data_preview"
response = requests.get(url)
with open("bus_delays.csv","wb") as file:
  file.write(response.content)
  print("Data is Ready")

Data is Ready


In [128]:
from google.colab import files
files.upload()

{}

In [129]:
import csv
file_name = "Bus_Breakdown_and_Delays_20260922.csv"
with open(file_name, "r", encoding="utf-8-sig",errors="replace", newline="") as f:
  reader = csv.reader(f)

  for i, row in enumerate(reader, start=1):
    if len(row) != 21:
        print("Problem row:", i)
        print("number of columns:", len(row))
        print(row)
        break
  else:
    print("All rows have 21 c")

All rows have 21 c


In [130]:
df = pl.read_csv(
    "Bus_Breakdown_and_Delays_20260922.csv",
    infer_schema_length=0
)
df.head()

School_Year,Busbreakdown_ID,Run_Type,Bus_No,Route_Number,Reason,Schools_Serviced,Occurred_On,Created_On,Boro,Bus_Company_Name,How_Long_Delayed,Number_Of_Students_On_The_Bus,Has_Contractor_Notified_Schools,Has_Contractor_Notified_Parents,Have_You_Alerted_OPT,Informed_On,Incident_Number,Last_Updated_On,Breakdown_or_Running_Late,School_Age_or_PreK
str,str,str,str,str,str,str,str,str,str,str,str,str,str,str,str,str,str,str,str,str
"""2015-2016""","""1272996""","""Special Ed AM Run""","""2256""","""K355""","""Won`t Start""","""15879""","""06/03/2016 06:03:00 AM""","""06/03/2016 06:05:00 AM""","""Brooklyn""","""RELIANT TRANS, INC. (B232""",null,"""0""","""Yes""","""Yes""","""Yes""","""06/03/2016 06:05:00 AM""",null,"""06/03/2016 06:05:05 AM""","""Breakdown""","""School-Age"""
"""2016-2017""","""1313907""","""Special Ed AM Run""","""4481""","""K002""","""Won`t Start""","""22001""","""01/10/2017 05:55:00 AM""","""01/10/2017 05:57:00 AM""","""Brooklyn""","""NEW DAWN TRANSIT, LLC (B2""","""30min""","""0""","""Yes""","""Yes""","""No""","""01/10/2017 05:57:00 AM""",null,"""01/10/2017 05:57:42 AM""","""Running Late""","""School-Age"""
"""2015-2016""","""1267391""","""Special Ed AM Run""","""GT2114""","""Y925""","""Won`t Start""","""11012""","""05/11/2016 05:50:00 AM""","""05/11/2016 05:59:00 AM""","""Bronx""","""DON THOMAS BUSES, INC. (B""","""30 MIN""","""0""","""Yes""","""Yes""","""Yes""","""05/11/2016 05:59:00 AM""",null,"""05/11/2016 05:59:42 AM""","""Running Late""","""School-Age"""
"""2016-2017""","""1312240""","""General Ed AM Run""","""2436""","""K1546""","""Other""","""15032""","""01/05/2017 06:00:00 AM""","""01/05/2017 06:05:00 AM""","""Brooklyn""","""ALL AMERICAN SCHOOL BUS C""","""20 MINS""","""0""","""No""","""No""","""No""","""01/05/2017 06:05:00 AM""",null,"""01/05/2017 06:05:48 AM""","""Running Late""","""School-Age"""
"""2016-2017""","""1312351""","""Special Ed AM Run""","""5814D""","""X441""","""Mechanical Problem""","""11002""","""01/05/2017 06:45:00 AM""","""01/05/2017 06:54:00 AM""","""Bronx""","""LITTLE LISA BUS CO. INC.""",null,"""0""","""Yes""","""Yes""","""No""","""01/05/2017 06:54:00 AM""",null,"""01/05/2017 10:50:02 AM""","""Breakdown""","""School-Age"""


In [131]:
df.shape


(1297681, 21)

In [132]:
df.columns

['School_Year',
 'Busbreakdown_ID',
 'Run_Type',
 'Bus_No',
 'Route_Number',
 'Reason',
 'Schools_Serviced',
 'Occurred_On',
 'Created_On',
 'Boro',
 'Bus_Company_Name',
 'How_Long_Delayed',
 'Number_Of_Students_On_The_Bus',
 'Has_Contractor_Notified_Schools',
 'Has_Contractor_Notified_Parents',
 'Have_You_Alerted_OPT',
 'Informed_On',
 'Incident_Number',
 'Last_Updated_On',
 'Breakdown_or_Running_Late',
 'School_Age_or_PreK']

In [133]:
df.schema

Schema([('School_Year', String),
        ('Busbreakdown_ID', String),
        ('Run_Type', String),
        ('Bus_No', String),
        ('Route_Number', String),
        ('Reason', String),
        ('Schools_Serviced', String),
        ('Occurred_On', String),
        ('Created_On', String),
        ('Boro', String),
        ('Bus_Company_Name', String),
        ('How_Long_Delayed', String),
        ('Number_Of_Students_On_The_Bus', String),
        ('Has_Contractor_Notified_Schools', String),
        ('Has_Contractor_Notified_Parents', String),
        ('Have_You_Alerted_OPT', String),
        ('Informed_On', String),
        ('Incident_Number', String),
        ('Last_Updated_On', String),
        ('Breakdown_or_Running_Late', String),
        ('School_Age_or_PreK', String)])

In [134]:
df.null_count()

School_Year,Busbreakdown_ID,Run_Type,Bus_No,Route_Number,Reason,Schools_Serviced,Occurred_On,Created_On,Boro,Bus_Company_Name,How_Long_Delayed,Number_Of_Students_On_The_Bus,Has_Contractor_Notified_Schools,Has_Contractor_Notified_Parents,Have_You_Alerted_OPT,Informed_On,Incident_Number,Last_Updated_On,Breakdown_or_Running_Late,School_Age_or_PreK
u32,u32,u32,u32,u32,u32,u32,u32,u32,u32,u32,u32,u32,u32,u32,u32,u32,u32,u32,u32,u32
1,0,31,57,134,144304,3625,1,1,24079,3492,179917,71,0,0,0,1,1283973,1,0,0


In [135]:
df.select(
    pl.col("Busbreakdown_ID").n_unique().alias("unique_ids")
)

unique_ids
u32
1297681


In [136]:
df.select([
    "Busbreakdown_ID",
    "Occurred_On",
    "Created_On",
    "Number_Of_Students_On_The_Bus",
    "How_Long_Delayed"
]).head(10)

Busbreakdown_ID,Occurred_On,Created_On,Number_Of_Students_On_The_Bus,How_Long_Delayed
str,str,str,str,str
"""1272996""","""06/03/2016 06:03:00 AM""","""06/03/2016 06:05:00 AM""","""0""",null
"""1313907""","""01/10/2017 05:55:00 AM""","""01/10/2017 05:57:00 AM""","""0""","""30min"""
"""1267391""","""05/11/2016 05:50:00 AM""","""05/11/2016 05:59:00 AM""","""0""","""30 MIN"""
"""1312240""","""01/05/2017 06:00:00 AM""","""01/05/2017 06:05:00 AM""","""0""","""20 MINS"""
"""1312351""","""01/05/2017 06:45:00 AM""","""01/05/2017 06:54:00 AM""","""0""",null
"""1320379""","""02/01/2017 02:17:00 PM""","""02/01/2017 02:24:00 PM""","""0""",null
"""1329633""","""03/15/2017 05:54:00 AM""","""03/15/2017 05:59:00 AM""","""0""",null
"""1339561""","""04/12/2017 02:00:00 PM""","""04/12/2017 02:14:00 PM""","""4""",null
"""1321074""","""02/06/2017 06:01:00 AM""","""02/06/2017 06:02:00 AM""","""0""","""15 MINS"""


In [137]:
df = df.with_columns(
    pl.col("Busbreakdown_ID").cast(pl.Int64)
)
df.schema["Busbreakdown_ID"]

Int64

In [138]:
df = df.with_columns(
    pl.col("Occurred_On")
    .str.strptime(
        pl.Datetime,format="%m/%d/%Y %I:%M:%S %p",
        strict=False
    )
    .alias("Occurred_On")
)
df.select("Occurred_On").head()

Occurred_On
datetime[μs]
2016-06-03 06:03:00
2017-01-10 05:55:00
2016-05-11 05:50:00
2017-01-05 06:00:00
2017-01-05 06:45:00


In [139]:
df = df.with_columns(
    pl.col("Created_On")
    .str.strptime(
        pl.Datetime,format="%m/%d/%Y %I:%M:%S %p",
        strict=False
    )
    .alias("Created_On")
)
df.select("Occurred_On", "Created_On").head()

Occurred_On,Created_On
datetime[μs],datetime[μs]
2016-06-03 06:03:00,2016-06-03 06:05:00
2017-01-10 05:55:00,2017-01-10 05:57:00
2016-05-11 05:50:00,2016-05-11 05:59:00
2017-01-05 06:00:00,2017-01-05 06:05:00
2017-01-05 06:45:00,2017-01-05 06:54:00


In [140]:
df.select(
    (
        pl.col("Number_Of_Students_On_The_Bus")
        .cast(pl.Int64, strict=False)
        .is_null()
        &
        pl.col("Number_Of_Students_On_The_Bus").is_not_null()
    )
    .sum()
    .alias("invalid_numeric_values")
)

invalid_numeric_values
u32
441


In [141]:
df.filter(
    pl.col("Number_Of_Students_On_The_Bus")
    .cast(pl.Int64, strict=False)
    .is_null()
    &
    pl.col("Number_Of_Students_On_The_Bus").is_not_null()
).select("Number_Of_Students_On_The_Bus"
).unique()

Number_Of_Students_On_The_Bus
str
"""4,113"""
"""2,408"""
"""1,472"""
"""1,382"""
"""1,835"""
…
"""1,254"""
"""1,330"""
"""1,188"""


In [142]:
df.select(
    (
        pl.col("Number_Of_Students_On_The_Bus")
        .str.replace_all(",", "")
        .cast(pl.Int64, strict=False)
        .is_null()
        &
        pl.col("Number_Of_Students_On_The_Bus").is_not_null()
    )
    .sum()
    .alias("still_invalid_after_removing_commas")
)

still_invalid_after_removing_commas
u32
0


In [143]:
df = df.with_columns(
    pl.col("Number_Of_Students_On_The_Bus")
    .str.replace_all(",", "")
    .cast(pl.Int64)
    .alias("Number_Of_Students_On_The_Bus")
)

df.select("Number_Of_Students_On_The_Bus").head()

Number_Of_Students_On_The_Bus
i64
0
0
0
0
0


In [144]:
df.select("How_Long_Delayed").drop_nulls().unique().head(30)

How_Long_Delayed
str
"""..55"""
"""2 HOURS TR"""
"""30 -40 MIN"""
"""1520 mins"""
""".120 min"""
…
"""30 min+"""
"""1-MORE HR"""
"""15 min"""


In [145]:
df.group_by("How_Long_Delayed")\
  .len()\
  .sort("len",descending=True)\
  .head(30)

How_Long_Delayed,len
str,u32
"""16-30 Min""",216049
null,179917
"""31-45 Min""",179640
"""46-60 Min""",100507
"""61-90 Min""",97846
…,…
"""30""",5370
"""15 MIN""",5205
"""45 MIN""",5087


In [146]:
df = df.with_columns(
    pl.col("How_Long_Delayed")
    .str.strip_chars()
    .str.to_uppercase()
    .str.replace_all(r"\s+","")
    .alias("How_Long_Delayed_Clean")
)

In [147]:
df.group_by("How_Long_Delayed_Clean")\
  .len() \
  .sort("len", descending=True) \
  .head(20)

How_Long_Delayed_Clean,len
str,u32
"""16-30MIN""",216049
null,179917
"""31-45MIN""",179640
"""46-60MIN""",100507
"""61-90MIN""",97846
…,…
"""10MIN""",10995
"""25MINS""",10641
"""30MINUTES""",10388


In [148]:
df = df.with_columns(
    pl.col("How_Long_Delayed_Clean")
    .str.replace_all(r"MINUTES|MINS","MIN")
    .alias("How_Long_Delayed_Clean")
)

In [149]:
df.group_by("How_Long_Delayed_Clean") \
  .len() \
  .sort("len", descending=True) \
  .head(20)

How_Long_Delayed_Clean,len
str,u32
"""16-30MIN""",216049
null,179917
"""31-45MIN""",179640
"""46-60MIN""",100507
"""61-90MIN""",97846
…,…
"""35MIN""",7358
"""15-20MIN""",7192
"""20-30MIN""",6041


In [150]:
df.group_by("How_Long_Delayed_Clean") \
  .len() \
  .sort("len", descending=True) \
  .head(20)

How_Long_Delayed_Clean,len
str,u32
"""16-30MIN""",216049
null,179917
"""31-45MIN""",179640
"""46-60MIN""",100507
"""61-90MIN""",97846
…,…
"""35MIN""",7358
"""15-20MIN""",7192
"""20-30MIN""",6041


In [151]:
df = df.with_columns(
    pl.col("How_Long_Delayed_Clean")
    .str.extract(r"^(\d+)-(\d+)MIN$", 1)
    .cast(pl.Int64, strict=False)
    .alias("Delay_Min_Minutes")
)

df.select([
    "How_Long_Delayed_Clean",
    "Delay_Min_Minutes"
]).filter(
    pl.col("Delay_Min_Minutes").is_not_null()
).head(10)

How_Long_Delayed_Clean,Delay_Min_Minutes
str,i64
"""16-30MIN""",16
"""46-60MIN""",46
"""31-45MIN""",31
"""16-30MIN""",16
"""16-30MIN""",16
"""46-60MIN""",46
"""16-30MIN""",16
"""16-30MIN""",16
"""16-30MIN""",16


In [152]:
df = df.with_columns(
    pl.col("How_Long_Delayed_Clean")
    .str.replace_all("–", "-")
    .str.replace_all("—", "-")
    .alias("How_Long_Delayed_Clean")
)

In [153]:
df = df.with_columns(
    pl.col("How_Long_Delayed_Clean")
    .str.extract(r"^(\d+)-(\d+)MIN$", 2)
    .cast(pl.Int64, strict=False)
    .alias("Delay_Max_Minutes")
)

df.select([
    "How_Long_Delayed_Clean",
    "Delay_Min_Minutes",
    "Delay_Max_Minutes"
]).filter(
    pl.col("Delay_Max_Minutes").is_not_null()
).head(10)

How_Long_Delayed_Clean,Delay_Min_Minutes,Delay_Max_Minutes
str,i64,i64
"""16-30MIN""",16,30
"""46-60MIN""",46,60
"""31-45MIN""",31,45
"""16-30MIN""",16,30
"""16-30MIN""",16,30
"""46-60MIN""",46,60
"""16-30MIN""",16,30
"""16-30MIN""",16,30
"""16-30MIN""",16,30


In [154]:
df = df.with_columns(
    pl.when(
        pl.col("How_Long_Delayed_Clean")
        .str.contains(r"^\d+MIN$")
    )
    .then(
        pl.col("How_Long_Delayed_Clean")
        .str.extract(r"^(\d+)MIN$", 1)
        .cast(pl.Int64)
    )
    .otherwise(pl.col("Delay_Min_Minutes"))
    .alias("Delay_Min_Minutes"),

    pl.when(
        pl.col("How_Long_Delayed_Clean")
        .str.contains(r"^\d+MIN$")
    )
    .then(
        pl.col("How_Long_Delayed_Clean")
        .str.extract(r"^(\d+)MIN$", 1)
        .cast(pl.Int64)
    )
    .otherwise(pl.col("Delay_Max_Minutes"))
    .alias("Delay_Max_Minutes")
)

In [155]:
df.select([
    "How_Long_Delayed_Clean",
    "Delay_Min_Minutes",
    "Delay_Max_Minutes"
]).filter(
    pl.col("How_Long_Delayed_Clean").str.contains(r"^\d+MIN$")
).head(10)

How_Long_Delayed_Clean,Delay_Min_Minutes,Delay_Max_Minutes
str,i64,i64
"""30MIN""",30,30
"""30MIN""",30,30
"""20MIN""",20,20
"""15MIN""",15,15
"""40MIN""",40,40
"""15MIN""",15,15
"""30MIN""",30,30
"""45MIN""",45,45
"""20MIN""",20,20


In [156]:
df = df.with_columns(
    pl.when(
        pl.col("How_Long_Delayed_Clean")
        .str.contains(r"^\d+$")
    )
    .then(
        pl.col("How_Long_Delayed_Clean")
        .cast(pl.Int64, strict=False)
    )
    .otherwise(pl.col("Delay_Min_Minutes"))
    .alias("Delay_Min_Minutes"),

    pl.when(
        pl.col("How_Long_Delayed_Clean")
        .str.contains(r"^\d+$")
    )
    .then(
        pl.col("How_Long_Delayed_Clean")
        .cast(pl.Int64, strict=False)
    )
    .otherwise(pl.col("Delay_Max_Minutes"))
    .alias("Delay_Max_Minutes")
)

In [157]:
df.select([
    "How_Long_Delayed_Clean",
    "Delay_Min_Minutes",
    "Delay_Max_Minutes"
]).filter(
    pl.col("How_Long_Delayed_Clean").str.contains(r"^\d+$")
).head(10)

How_Long_Delayed_Clean,Delay_Min_Minutes,Delay_Max_Minutes
str,i64,i64
"""25""",25,25
"""15""",15,15
"""0""",0,0
"""30""",30,30
"""30""",30,30
"""20""",20,20
"""45""",45,45
"""5""",5,5
"""20""",20,20


In [158]:
df = df.with_columns(
    pl.when(
        pl.col("How_Long_Delayed_Clean")
        .str.contains(r"^\d+HR$")
    )
    .then(
        pl.col("How_Long_Delayed_Clean")
        .str.extract(r"^(\d+)HR$", 1)
        .cast(pl.Int64, strict=False) * 60
    )
    .otherwise(pl.col("Delay_Min_Minutes"))
    .alias("Delay_Min_Minutes"),

    pl.when(
        pl.col("How_Long_Delayed_Clean")
        .str.contains(r"^\d+HR$")
    )
    .then(
        pl.col("How_Long_Delayed_Clean")
        .str.extract(r"^(\d+)HR$", 1)
        .cast(pl.Int64, strict=False) * 60
    )
    .otherwise(pl.col("Delay_Max_Minutes"))
    .alias("Delay_Max_Minutes")
)

In [159]:
df.select([
    "How_Long_Delayed_Clean",
    "Delay_Min_Minutes",
    "Delay_Max_Minutes"
]).filter(
    pl.col("How_Long_Delayed_Clean").str.contains(r"^\d+HR$")
).head(10)

How_Long_Delayed_Clean,Delay_Min_Minutes,Delay_Max_Minutes
str,i64,i64
"""1HR""",60,60
"""1HR""",60,60
"""1HR""",60,60
"""1HR""",60,60
"""1HR""",60,60
"""1HR""",60,60
"""1HR""",60,60
"""1HR""",60,60
"""1HR""",60,60


In [160]:
df =df.with_columns(
    (
        (pl.col("Delay_Min_Minutes") + pl.col("Delay_Max_Minutes")) / 2
    )
    .round(0)
    .cast(pl.Int64)
    .alias("Delay_Avg_Minutes")
)

In [161]:
df.select([
    "How_Long_Delayed_Clean",
    "Delay_Min_Minutes",
    "Delay_Max_Minutes",
    "Delay_Avg_Minutes"
]).head(10)

How_Long_Delayed_Clean,Delay_Min_Minutes,Delay_Max_Minutes,Delay_Avg_Minutes
str,i64,i64,i64
null,null,null,null
"""30MIN""",30,30,30
"""30MIN""",30,30,30
"""20MIN""",20,20,20
null,null,null,null
null,null,null,null
null,null,null,null
null,null,null,null
"""15MIN""",15,15,15


In [162]:
df = df.with_columns(
    pl.col("Informed_On")
    .str.strptime(
        pl.Datetime,
        format="%m/%d/%Y %I:%M:%S %p",
        strict=False
    )
    .alias("Informed_On"),

    pl.col("Last_Updated_On")
    .str.strptime(
        pl.Datetime,
        format="%m/%d/%Y %I:%M:%S %p",
        strict=False
    )
    .alias("Last_Updated_On")
)

In [163]:
df.select([
    "Occurred_On",
    "Created_On",
    "Informed_On",
    "Last_Updated_On"
]).head()

Occurred_On,Created_On,Informed_On,Last_Updated_On
datetime[μs],datetime[μs],datetime[μs],datetime[μs]
2016-06-03 06:03:00,2016-06-03 06:05:00,2016-06-03 06:05:00,2016-06-03 06:05:05
2017-01-10 05:55:00,2017-01-10 05:57:00,2017-01-10 05:57:00,2017-01-10 05:57:42
2016-05-11 05:50:00,2016-05-11 05:59:00,2016-05-11 05:59:00,2016-05-11 05:59:42
2017-01-05 06:00:00,2017-01-05 06:05:00,2017-01-05 06:05:00,2017-01-05 06:05:48
2017-01-05 06:45:00,2017-01-05 06:54:00,2017-01-05 06:54:00,2017-01-05 10:50:02


In [164]:
string_cols = [
    col for col, dtype in df.schema.items()
    if dtype == pl.String
]

df = df.with_columns(
    [
        pl.col(col)
        .str.strip_chars()
        .alias(col)
        for col in string_cols
    ]
)

print("Cleaned string columns:", len(string_cols))

Cleaned string columns: 16


In [165]:
df = df.with_columns([
    pl.col("Occurred_On").dt.year().alias("Incident_Year"),
    pl.col("Occurred_On").dt.month().alias("Incident_Month"),
    pl.col("Occurred_On").dt.day().alias("Incident_Day"),
    pl.col("Occurred_On").dt.hour().alias("Incident_Hour"),
    pl.col("Occurred_On").dt.weekday().alias("Incident_Weekday")
])

In [166]:
df.select([
    "Occurred_On",
    "Incident_Year",
    "Incident_Month",
    "Incident_Day",
    "Incident_Hour",
    "Incident_Weekday"
]).head()

Occurred_On,Incident_Year,Incident_Month,Incident_Day,Incident_Hour,Incident_Weekday
datetime[μs],i32,i8,i8,i8,i8
2016-06-03 06:03:00,2016,6,3,6,5
2017-01-10 05:55:00,2017,1,10,5,2
2016-05-11 05:50:00,2016,5,11,5,3
2017-01-05 06:00:00,2017,1,5,6,4
2017-01-05 06:45:00,2017,1,5,6,4


In [167]:
df.select([
    pl.len().alias("total_rows"),
    pl.col("Busbreakdown_ID").n_unique().alias("unique_busbreakdown_ids"),
    pl.col("Occurred_On").is_null().sum().alias("null_occurred_on"),
    pl.col("Number_Of_Students_On_The_Bus").is_null().sum().alias("null_student_count"),
    pl.col("Delay_Avg_Minutes").is_null().sum().alias("null_delay_minutes")
])

total_rows,unique_busbreakdown_ids,null_occurred_on,null_student_count,null_delay_minutes
u32,u32,u32,u32,u32
1297681,1297681,1,71,241879


In [168]:
df = df.with_columns(
    pl.when(
        pl.col("Occurred_On").is_null() |
        pl.col("Created_On").is_null()
    )
    .then(pl.lit("MISSING_DATE"))

    .when(
        pl.col("Occurred_On") <= pl.col("Created_On")
    )
    .then(pl.lit("VALID"))

    .when(
        pl.col("Occurred_On").dt.date() ==
        pl.col("Created_On").dt.date()
    )
    .then(pl.lit("SAME_DAY_ORDER_ISSUE"))

    .otherwise(pl.lit("DATE_SEQUENCE_ISSUE"))
    .alias("Date_Quality_Flag")
)

df.group_by("Date_Quality_Flag").len().sort("len", descending=True)

Date_Quality_Flag,len
str,u32
"""VALID""",1268884
"""SAME_DAY_ORDER_ISSUE""",28731
"""DATE_SEQUENCE_ISSUE""",65
"""MISSING_DATE""",1


In [169]:
# Overwrite the Parquet file with the final cleaned DataFrame
df.write_parquet(
    "bus_delays_clean.parquet",
    compression="snappy"
)

print("Final Parquet file updated")

Final Parquet file updated


In [170]:
from google.colab import auth
auth.authenticate_user()

In [171]:
!gcloud projects list --format="table(projectId,name)"

PROJECT_ID             NAME
electric-nomad-388815  My First Project


In [172]:
from google.cloud import bigquery

project_id = "electric-nomad-388815"

client = bigquery.Client(project=project_id)

print("Connected to BigQuery project:", project_id)

Connected to BigQuery project: electric-nomad-388815


In [173]:
dataset_id = f"{project_id}.transportation_data"

dataset = bigquery.Dataset(dataset_id)
dataset.location = "US"

dataset = client.create_dataset(dataset, exists_ok=True)

print("Dataset ready:", dataset_id)

Dataset ready: electric-nomad-388815.transportation_data


In [174]:
table_id = f"{project_id}.transportation_data.bus_delays"

job_config = bigquery.LoadJobConfig(
    source_format=bigquery.SourceFormat.PARQUET,
    write_disposition=bigquery.WriteDisposition.WRITE_TRUNCATE
)

with open("bus_delays_clean.parquet", "rb") as source_file:
    load_job = client.load_table_from_file(
        source_file,
        table_id,
        job_config=job_config
    )

load_job.result()

print("Parquet loaded successfully into BigQuery")

Parquet loaded successfully into BigQuery


In [175]:
table = client.get_table(table_id)
print("Rows in BigQuery:", table.num_rows)
print("Columns in BigQuery:", len(table.schema))

Rows in BigQuery: 1297681
Columns in BigQuery: 31


In [176]:
raw_check = pl.read_csv(
    "Bus_Breakdown_and_Delays_20260922.csv",
    infer_schema_length=0
)

print("Raw CSV rows:", raw_check.height)
print("Current cleaned df rows:", df.height)

Raw CSV rows: 1297681
Current cleaned df rows: 1297681


In [177]:
query = f"""
SELECT COUNT(*) AS total_rows
from {table_id}
"""
result = client.query(query).to_dataframe()
result

,total_rows
0,1297681


In [178]:
query = f"""
SELECT
    Reason,
    COUNT(*) AS total_incidents
FROM `{table_id}`
WHERE Reason IS NOT NULL
GROUP BY Reason
ORDER BY total_incidents DESC
LIMIT 10
"""

result = client.query(query).to_dataframe()

result

,Reason,total_incidents
0,Heavy Traffic,683361
1,Other,192926
2,Mechanical Problem,99776
3,Problem Run,42382
4,Won`t Start,38053
5,Weather Conditions,35600
6,Flat Tire,22635
7,Late return from Field Trip,17369
8,Delayed by School,11346
9,Accident,9929


In [179]:
query = f"""
SELECT
    Boro,
    COUNT(*) AS total_incidents
FROM `{table_id}`
WHERE Boro IS NOT NULL
GROUP BY Boro
ORDER BY total_incidents DESC
"""

borough_result = client.query(query).to_dataframe()

borough_result

,Boro,total_incidents
0,Brooklyn,316491
1,Bronx,294986
2,Manhattan,256689
3,Queens,249499
4,Staten Island,62311
5,Nassau County,43737
6,Westchester,33110
7,New Jersey,6157
8,Connecticut,4445
9,All Boroughs,3406


In [180]:
query = f"""
SELECT
    Incident_Hour,
    COUNT(*) AS total_incidents
FROM `{table_id}`
WHERE Incident_Hour IS NOT NULL
GROUP BY Incident_Hour
ORDER BY Incident_Hour
"""

hour_result = client.query(query).to_dataframe()

hour_result

,Incident_Hour,total_incidents
0,0,4820
1,1,1619
2,2,2889
3,3,1161
4,4,710
5,5,38301
6,6,364006
7,7,470163
8,8,130157
9,9,13301


In [181]:
query = f"""
SELECT
    Reason,
    ROUND(AVG(Delay_Avg_Minutes), 2) AS avg_delay_minutes,
    COUNT(*) AS total_incidents
FROM `{table_id}`
WHERE
    Reason IS NOT NULL
    AND Delay_Avg_Minutes IS NOT NULL
GROUP BY Reason
HAVING COUNT(*) >= 100
ORDER BY avg_delay_minutes DESC
LIMIT 10
"""

delay_reason_result = client.query(query).to_dataframe()

delay_reason_result

,Reason,avg_delay_minutes,total_incidents
0,Late return from Field Trip,46.60,15323
1,Accident,45.39,7821
2,Mechanical Problem,42.43,40535
3,Won`t Start,38.63,10471
4,Delayed by School,38.34,9902
5,Weather Conditions,38.25,31661
6,Flat Tire,38.14,9563
7,Other,37.74,168520
8,Problem Run,37.45,40881
9,Heavy Traffic,34.68,648642


In [182]:
query = f"""
SELECT
    Incident_Year,
    COUNT(*) AS total_incidents
FROM `{table_id}`
WHERE Incident_Year IS NOT NULL
GROUP BY Incident_Year
ORDER BY Incident_Year
"""

year_result = client.query(query).to_dataframe()

year_result

,Incident_Year,total_incidents
0,1900,2
1,1979,4
2,1983,1
3,1999,1002
4,2000,2560
5,2001,1205
6,2002,4356
7,2003,5482
8,2004,7580
9,2005,9455


In [183]:
query = f"""
SELECT
    School_Year,
    Occurred_On,
    Incident_Year,
    Busbreakdown_ID,
    Reason
FROM `{table_id}`
WHERE Incident_Year < 2015
   OR Incident_Year > 2026
ORDER BY Incident_Year
LIMIT 30
"""

date_quality_check = client.query(query).to_dataframe()

date_quality_check

,School_Year,Occurred_On,Incident_Year,Busbreakdown_ID,Reason
0,1899-1900,1900-01-01 00:00:00+00:00,1900,780993,None
1,1899-1900,1900-03-14 00:00:00+00:00,1900,783677,None
2,1979-1980,1979-11-08 00:00:00+00:00,1979,783479,None
3,1979-1980,1979-10-22 00:00:00+00:00,1979,783487,None
4,1979-1980,1979-10-04 00:00:00+00:00,1979,783591,None
5,1979-1980,1979-10-21 00:00:00+00:00,1979,781877,None
6,1983-1984,1983-09-18 16:15:00+00:00,1983,1146576,Mechanical Problem
7,1999-2000,1999-09-17 00:00:00+00:00,1999,780883,None
8,1999-2000,1999-09-13 00:00:00+00:00,1999,780881,None
9,1999-2000,1999-10-15 00:00:00+00:00,1999,780885,None


In [184]:
query = f"""
SELECT
    School_Year,
    Busbreakdown_ID,
    Occurred_On,
    Created_On,
    Last_Updated_On,
    Reason,
    Boro
FROM `{table_id}`
WHERE Incident_Year = 2027
"""

future_date_check = client.query(query).to_dataframe()

future_date_check

,School_Year,Busbreakdown_ID,Occurred_On,Created_On,Last_Updated_On,Reason,Boro
0,2026-2027,827680,2027-02-26 14:00:00+00:00,2007-02-26 14:18:00+00:00,2007-02-27 08:05:48+00:00,None,Manhattan


In [185]:
query = f"""
SELECT
    COUNT(*) AS invalid_date_sequence
FROM `{table_id}`
WHERE Occurred_On > Created_On
"""

date_issue_count = client.query(query).to_dataframe()

date_issue_count

,invalid_date_sequence
0,28796


In [186]:
query = f"""
SELECT
    CASE
        WHEN DATE_DIFF(DATE(Occurred_On), DATE(Created_On), DAY) = 0
            THEN 'Same day'
        WHEN DATE_DIFF(DATE(Occurred_On), DATE(Created_On), DAY) BETWEEN 1 AND 7
            THEN '1-7 days'
        WHEN DATE_DIFF(DATE(Occurred_On), DATE(Created_On), DAY) BETWEEN 8 AND 365
            THEN '8-365 days'
        ELSE 'More than 1 year'
    END AS date_gap,
    COUNT(*) AS records
FROM `{table_id}`
WHERE Occurred_On > Created_On
GROUP BY date_gap
ORDER BY records DESC
"""

gap_result = client.query(query).to_dataframe()

gap_result


,date_gap,records
0,Same day,28731
1,1-7 days,43
2,More than 1 year,12
3,8-365 days,10


In [187]:
query = f"""
SELECT
    Date_Quality_Flag,
    COUNT(*) AS records
FROM `{table_id}`
GROUP BY Date_Quality_Flag
ORDER BY records DESC
"""

quality_result = client.query(query).to_dataframe()

quality_result

,Date_Quality_Flag,records
0,VALID,1268884
1,SAME_DAY_ORDER_ISSUE,28731
2,DATE_SEQUENCE_ISSUE,65
3,MISSING_DATE,1


## Project Summary

Built an end-to-end ETL pipeline using NYC Bus Breakdown and Delays open data.

### Pipeline
CSV → Polars → Parquet → Google BigQuery → SQL Analytics

### Key Work Completed
- Processed 1,297,681 records using Polars.
- Cleaned inconsistent text and numeric fields.
- Converted timestamp columns to proper datetime types.
- Standardized delay-duration values and created numeric delay metrics.
- Created derived date features for analysis.
- Added data-quality validation for date inconsistencies.
- Converted the cleaned dataset to Parquet using Snappy compression.
- Loaded the final dataset into Google BigQuery.
- Validated 1,297,681 rows and 31 columns in BigQuery.
- Performed SQL analysis on delay reasons, boroughs, incident hours, delay duration, and yearly trends.

### Data Quality Results
- VALID: 1,268,884
- SAME_DAY_ORDER_ISSUE: 28,731
- DATE_SEQUENCE_ISSUE: 65
- MISSING_DATE: 1

### Technologies
Python, Polars, Parquet, Google Colab, Google BigQuery, SQL